django/db/models/expressions.py

```diff
@@ -1768,6 +1769,14 @@ def set_source_expressions(self, exprs):
     def _resolve_output_field(self):
         return self.query.output_field
 
+    def resolve_expression(self, *args, **kwargs):
+        resolved = super().resolve_expression(*args, **kwargs)
+        if type(self) is Subquery and self.template == Subquery.template:
+            # Subquery is an unnecessary shim for a resolved query as it
+            # complexifies the lookup's right-hand-side introspection.
+            return resolved.query
```

Comment:
@charettes do you think I need something like this? If so I'll add a test.
```suggestion
            if self.output_field != resolved.query.output_field:
                resolved.query._output_field = self.output_field
            return resolved.query
```

Reply:
I cloned an existing test using `Subquery(..., output_field=...`) in a non-composite-pk situation, and changed the `output_field` from `IntegerField()` to `DecimalField()` and pushed it to this PR.

Before I caught the issue, I failed like this (but passes on main):
```
FAIL: test_annotation_with_outerref_and_output_field (expressions.tests.BasicExpressionsTests.test_annotation_with_outerref_and_output_field)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/.../django/tests/expressions/tests.py", line 973, in test_annotation_with_outerref_and_output_field
    self.assertIsInstance(gmbh_salary.max_ceo_salary_raise, Decimal)
    ~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: 2332 is not an instance of <class 'decimal.Decimal'>```